piwheels/master/the_scribe.py

```diff
@@ -420,6 +519,91 @@ def delete_version(self, package, version):
+        excluding any versions in *exclude*.
+
+        :param str package:
+            The name of the package to get versions for.
+
+        :type exclude: set or None
+        :param exclude:
+            Set of versions to exclude from results.
+        """
+        versions = [
+            row._replace(version=parse_version(row.version))
+            for row in self.db.get_project_versions(package)
+            if row.version not in exclude
+        ]
+        return sorted(versions, key=attrgetter('version'), reverse=True)
+
+    def get_files(self, package, exclude):
+        """
+        Retrieves a list of files and associated data for *package* excluding
+        any files from versions in *exclude*.
+
+        :param str package:
+            The name of the package to get versions for.
+
+        :type exclude: set or None
+        :param exclude:
+            Set of versions to exclude from results.
+        """
+        return [
+            row._replace(version=parse_version(row.version))
+            for row in self.db.get_project_files(package)
+            if row.version not in exclude
+        ]
+
+    def get_releases(self, versions, files):
+        """
+        Combines the given *files* and *versions* for a project into a list of
+        releases containing all the release information required to write out
+        the project page and project JSON file for the given package. Returns
+        the list of releases.
+
+        :param list versions:
+            The name of the package to get versions for.
+
+        :param list files:
+            Set of versions to exclude from results.
+        """
+        releases = []
+        for version in versions:
+            version_files = [
+                {
+                    'filename': f.filename,
+                    'filehash': f.filehash,
+                    'filesize': f.filesize,
+                    'builder_abi': f.builder_abi,
+                    'file_abi_tag': f.file_abi_tag,
+                    'platform': f.platform_tag,
+                    'apt_dependencies': f.dependencies,
+                }
+                for f in files
+                if f.version.original == version.version.original
+            ]
+            builds_succeeded = [abi for abi in version.builds_succeeded.split(',') if abi]
+            builds_failed = [abi for abi in version.builds_failed.split(',') if abi]
+            release = ProjectRelease(
+                version=version.version,
+                yanked=version.yanked,
+                released=version.released,
+                skip=version.skip,
+                stretch=_get_abi_status(
+                    'cp35m', builds_succeeded, builds_failed, version_files,
+                    version.skip),
+                buster=_get_abi_status(
+                    'cp37m', builds_succeeded, builds_failed, version_files,
+                    version.skip),
```

Comment:
Urggggggh - hard-coded ABIs *and* release names. Surely there's a better way of doing this? I wouldn't mind this sort of stuff in the *templates* as ultimately they probably want to be configurable for an installation anyway, but having this hard-coded in the master itself .... urgh.